src/Microsoft.AspNet.OData.Shared/Extensions/ODataPathExtensions.cs

```diff
@@ -11,46 +11,75 @@
+        /// <param name="path"><see cref="ODataPath"/>.</param>
+        /// <returns>Dictionary of keys.</returns>
         public static Dictionary<string, object> GetKeys(this ODataPath path)
         {
             if (path == null)
             {
                 throw Error.ArgumentNull(nameof(path));
             }
 
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
+            int position = path.Count - 1;
+            ODataPathSegment pathSegment = pathSegments[position];
 
-            // Books(1)/Authors(1000)/Namespace.SpecialAuthor
-            if (path.LastSegment is TypeSegment)
+            while (!(pathSegment is KeySegment) && position >= 0)
             {
-                ODataPath pathWithoutLastSegmentCastType = path.TrimEndingTypeSegment();
+                pathSegment = pathSegments[--position];
+            }
+
+            return ODataPathHelper.GetKeysFromKeySegment(pathSegment as KeySegment);
+        }
 
-                if (pathWithoutLastSegmentCastType.LastSegment is KeySegment)
-                {
-                    keys = GetKeysFromKeySegment(pathWithoutLastSegmentCastType.LastSegment as KeySegment);
-                }
+        /// <summary>
+        /// Return the last segment in the path, which is not a <see cref="TypeSegment"/> or <see cref="KeySegment"/>.
+        /// </summary>
+        /// <param name="path">The <see cref="ODataPath"/>.</param>
+        /// <returns>An <see cref="ODataPathSegment"/>.</returns>
+        public static ODataPathSegment GetLastNonTypeNonKeySegment(this ODataPath path)
+        {
+            if (path == null)
+            {
+                throw Error.ArgumentNull(nameof(path));
             }
-            // Books(1)/Authors/Namespace.SpecialAuthor/(1000)
-            else if (path.LastSegment is KeySegment)
+
+            if (path.Count == 1)
             {
-                keys = GetKeysFromKeySegment(path.LastSegment as KeySegment);
+                return path.LastSegment;
             }
 
-            return keys;
-        }
+            // If the path is Employees(2)/NewFriends(2)/Namespace.MyNewFriend where Namespace.MyNewFriend is a type segment,
+            // This method will return NewFriends NavigationPropertySegment.
 
-        private static Dictionary<string, object> GetKeysFromKeySegment(KeySegment keySegment)
-        {
-            Dictionary<string, object> keys = new Dictionary<string, object>();
+            List<ODataPathSegment> pathSegments = path.AsList();
+            int position = path.Count - 1;
+            ODataPathSegment pathSegment = pathSegments[position];
 
-            foreach (KeyValuePair<string, object> kvp in keySegment.Keys)
+            while ((pathSegment is TypeSegment || pathSegment is KeySegment) && position>=0)
             {
-                keys.Add(kvp.Key, kvp.Value);
+                pathSegment = pathSegments[--position];
```

Comment:
Can this potentially use a negative index? `pathSegments[-1]`? Since the while condition allows position to get to 0 at this point.

Reply:
Fixed it in the latest commit

Reply:
@habbes `AsList` is defined as an internal extension method